probe/endpoint/procspy/background_reader_linux.go

```diff
@@ -0,0 +1,130 @@
+package procspy
+
+import (
+	"bytes"
+	"io"
+	"math"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/weaveworks/scope/probe/process"
+)
+
+const (
+	initialRateLimitPeriod = 50 * time.Millisecond  // Read 20 * fdBlockSize file descriptors (/proc/PID/fd/*) per namespace per second
+	maxRateLimitPeriod     = 500 * time.Millisecond // Read at least 2 * fdBlockSize file descriptors per namespace per second
+	fdBlockSize            = uint64(300)            // Maximum number of /proc/PID/fd/* files to stat per rate-limit period
+	// (as a rule of thumb going through each block should be more expensive than reading /proc/PID/tcp{,6})
+	targetWalkTime = 10 * time.Second // Aim at walking all files in 10 seconds
+)
+
+type backgroundReader struct {
+	stopc         chan struct{}
+	mtx           sync.Mutex
+	latestBuf     *bytes.Buffer
+	latestSockets map[uint64]*Proc
+}
+
+// starts a rate-limited background goroutine to read the expensive files from
+// proc.
+func newBackgroundReader(walker process.Walker) *backgroundReader {
+	br := &backgroundReader{
+		latestBuf: bytes.NewBuffer(make([]byte, 0, 5000)),
+		stopc:     make(chan struct{}),
+	}
+	go br.loop(walker)
+	return br
+}
+
+func (br *backgroundReader) stop() {
+	close(br.stopc)
+}
+
+func (br *backgroundReader) getWalkedProcPid(buf *bytes.Buffer) (map[uint64]*Proc, error) {
+	br.mtx.Lock()
+	defer br.mtx.Unlock()
+
+	_, err := io.Copy(buf, br.latestBuf)
+
+	return br.latestSockets, err
+}
+
+func (br *backgroundReader) loop(walker process.Walker) {
+	var (
+		begin           time.Time                      // when we started the last performWalk
+		tickc           = time.After(time.Millisecond) // fire immediately
+		walkc           chan map[uint64]*Proc          // initially nil, i.e. off
+		walkBuf         = bytes.NewBuffer(make([]byte, 0, 5000))
```

Comment:
I think there is no need to preallocate a buffer here. Whatever allocation costs exist will be amortized very quickly.